File: dotnet/backend/HOMSys.Infrastructure/Data/Dbf/DbfReader.cs
using System.Text;

namespace HOMSys.Infrastructure.Data.Dbf;

/// <summary>
/// Minimal reader for Visual FoxPro DBF files.
///
/// Only what the reference-data import needs: header, field descriptors, and
/// records with C / N / D / L / I / T fields. Memo (M) fields are NOT supported —
/// none of the tables we import have an .FPT, verified during staging.
///
/// Production DBFs are frequently held open by BMS, so the file is opened with
/// FileShare.ReadWrite. A plain File.OpenRead throws "being used by another
/// process" on a live share.
/// </summary>
public sealed class DbfReader : IDisposable
{
    private readonly FileStream _fs;
    private readonly Encoding _encoding;

    public IReadOnlyList<DbfField> Fields { get; }
    public int RecordCount { get; }
    public int RecordLength { get; }
    public int HeaderLength { get; }

    static DbfReader()
    {
        // DBF text is codepage 1252, which .NET Core does not register by default.
        Encoding.RegisterProvider(CodePagesEncodingProvider.Instance);
    }

    public DbfReader(string path)
    {
        _fs = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
        _encoding = Encoding.GetEncoding(1252);

        var header = new byte[32];
        ReadExact(header, 32);

        RecordCount = BitConverter.ToInt32(header, 4);
        HeaderLength = BitConverter.ToInt16(header, 8);
        RecordLength = BitConverter.ToInt16(header, 10);

        var fields = new List<DbfField>();
        var offset = 1;   // byte 0 of each record is the deleted flag
        var descriptor = new byte[32];

        while (_fs.Position < HeaderLength - 1)
        {
            ReadExact(descriptor, 32);
            if (descriptor[0] == 0x0D) break;   // header terminator

            var name = _encoding.GetString(descriptor, 0, 11).TrimEnd('\0').Trim();
            if (name.Length == 0) break;

            var field = new DbfField(
                name,
                (char)descriptor[11],
                descriptor[16],
                descriptor[17],
                offset);

            fields.Add(field);
            offset += field.Length;
        }

        Fields = fields;
    }

    /// <summary>Non-deleted records, streamed.</summary>
    public IEnumerable<DbfRecord> Records()
    {
        var buffer = new byte[RecordLength];

        for (var i = 0; i < RecordCount; i++)
        {
            _fs.Seek(HeaderLength + ((long)i * RecordLength), SeekOrigin.Begin);

            var read = 0;
            while (read < RecordLength)
            {
                var n = _fs.Read(buffer, read, RecordLength - read);
                if (n <= 0) break;
                read += n;
            }
            if (read < RecordLength) yield break;   // truncated tail

            if (buffer[0] == 0x2A) continue;        // deleted

            yield return new DbfRecord(buffer, Fields, _encoding) { RecNo = i + 1 };
        }
    }

    private void ReadExact(byte[] buffer, int count)
    {
        var read = 0;
        while (read < count)
        {
            var n = _fs.Read(buffer, read, count - read);
            if (n <= 0) throw new EndOfStreamException("Unexpected end of DBF header.");
            read += n;
        }
    }

    public void Dispose() => _fs.Dispose();
}

public sealed record DbfField(string Name, char Type, int Length, int Decimals, int Offset);

public sealed class DbfRecord(byte[] buffer, IReadOnlyList<DbfField> fields, Encoding encoding)
{
    /// <summary>1-based physical record position, matching VFP's RECNO() — stable
    /// identity for a DBF row across syncs, used to diff without relying on
    /// natural keys that can legitimately repeat (e.g. duplicate EFF_DATE rows).</summary>
    public int RecNo { get; init; }

    private DbfField? Find(string name) =>
        fields.FirstOrDefault(f => string.Equals(f.Name, name, StringComparison.OrdinalIgnoreCase));

    public string GetString(string name)
    {
        var f = Find(name);
        if (f is null) return string.Empty;
        return encoding.GetString(buffer, f.Offset, f.Length).Trim();
    }

    public int GetInt(string name)
    {
        var raw = GetString(name);
        return int.TryParse(raw, out var v) ? v : 0;
    }

    public decimal GetDecimal(string name)
    {
        var raw = GetString(name);
        return decimal.TryParse(raw, out var v) ? v : 0m;
    }

    /// <summary>Same as GetDecimal, but returns null for a blank field instead of 0.</summary>
    public decimal? GetDecimalOrNull(string name)
    {
        var raw = GetString(name);
        if (raw.Length == 0) return null;
        return decimal.TryParse(raw, out var v) ? v : null;
    }

    /// <summary>DBF date fields are 8 chars, YYYYMMDD. Blank means null.</summary>
    public DateOnly? GetDate(string name)
    {
        var raw = GetString(name);
        if (raw.Length != 8) return null;

        return int.TryParse(raw[..4], out var y) &&
               int.TryParse(raw.Substring(4, 2), out var m) &&
               int.TryParse(raw.Substring(6, 2), out var d) &&
               y is >= 1 and <= 9999 && m is >= 1 and <= 12 && d is >= 1 and <= 31
            ? new DateOnly(y, m, d)
            : null;
    }

    public bool GetBool(string name)
    {
        var raw = GetString(name);
        return raw.Length > 0 && (raw[0] is 'T' or 't' or 'Y' or 'y');
    }
}

File: watcher/dbf_reader.py
"""Minimal Visual FoxPro DBF reader — Python port of
HOMSys.Infrastructure\\Data\\Dbf\\DbfReader.cs, byte-for-byte compatible so
RecNo and parsed values match what's already in SQL.

Every field is read as text (cp1252, stripped) regardless of its declared DBF
type, then parsed on demand by GetInt/GetDecimal/GetDate/GetBool — same as
the C# reader. Deliberate divergence: get_date() returns None for an invalid
calendar date (e.g. Feb 30) instead of raising, unlike the C# code's latent
crash there.

Production DBFs are frequently held open by BMS; opened read-only, which on
Windows allows concurrent readers/writers same as the C# FileShare.ReadWrite.
"""
from __future__ import annotations

import struct
from dataclasses import dataclass
from datetime import date


@dataclass
class DbfField:
    name: str
    type: str
    length: int
    decimals: int
    offset: int


class DbfRecord:
    __slots__ = ("buffer", "fields", "recno")

    def __init__(self, buffer: bytes, fields: list[DbfField], recno: int):
        self.buffer = buffer
        self.fields = fields
        self.recno = recno

    def _find(self, name: str) -> DbfField | None:
        for f in self.fields:
            if f.name.lower() == name.lower():
                return f
        return None

    def get_string(self, name: str) -> str:
        f = self._find(name)
        if f is None:
            return ""
        raw = self.buffer[f.offset:f.offset + f.length]
        return raw.decode("cp1252", errors="replace").strip()

    def get_int(self, name: str) -> int:
        raw = self.get_string(name)
        try:
            return int(raw)
        except ValueError:
            return 0

    def get_decimal(self, name: str) -> float:
        raw = self.get_string(name)
        try:
            return float(raw)
        except ValueError:
            return 0.0

    def get_decimal_or_null(self, name: str) -> float | None:
        raw = self.get_string(name)
        if raw == "":
            return None
        try:
            return float(raw)
        except ValueError:
            return None

    def get_date(self, name: str) -> str | None:
        """8-char YYYYMMDD -> ISO 'YYYY-MM-DD', or None if blank/invalid."""
        raw = self.get_string(name)
        if len(raw) != 8:
            return None
        try:
            y, m, d = int(raw[0:4]), int(raw[4:6]), int(raw[6:8])
        except ValueError:
            return None
        if not (1 <= y <= 9999 and 1 <= m <= 12 and 1 <= d <= 31):
            return None
        try:
            return date(y, m, d).isoformat()
        except ValueError:
            return None

    def get_bool(self, name: str) -> bool:
        raw = self.get_string(name)
        return len(raw) > 0 and raw[0] in "TtYy"


class DbfReader:
    def __init__(self, path: str):
        self.path = path
        self._f = open(path, "rb")

        header = self._read_exact(32)
        self.record_count = struct.unpack_from("<i", header, 4)[0]
        self.header_length = struct.unpack_from("<h", header, 8)[0]
        self.record_length = struct.unpack_from("<h", header, 10)[0]

        fields: list[DbfField] = []
        offset = 1  # byte 0 of each record is the deleted flag

        while self._f.tell() < self.header_length - 1:
            descriptor = self._read_exact(32)
            if descriptor[0] == 0x0D:  # header terminator
                break

            name = descriptor[0:11].decode("cp1252", errors="replace").rstrip("\x00").strip()
            if name == "":
                break

            length = descriptor[16]
            decimals = descriptor[17]
            fields.append(DbfField(name, chr(descriptor[11]), length, decimals, offset))
            offset += length

        self.fields = fields

    def records(self):
        """Non-deleted records, streamed."""
        for i in range(self.record_count):
            self._f.seek(self.header_length + i * self.record_length)
            buffer = self._f.read(self.record_length)
            if len(buffer) < self.record_length:
                return  # truncated tail

            if buffer[0] == 0x2A:  # deleted
                continue

            yield DbfRecord(buffer, self.fields, i + 1)

    def _read_exact(self, count: int) -> bytes:
        data = self._f.read(count)
        if len(data) < count:
            raise EOFError("Unexpected end of DBF header.")
        return data

    def close(self):
        self._f.close()

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc, tb):
        self.close()
